In [1]:
#Import the packages
import pandas as mypd
from sklearn import tree
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as myplot
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import math as mymath
import seaborn as mysb
from sklearn.ensemble import RandomForestClassifier
import random
from statsmodels.formula.api import ols
from scipy import stats
import statsmodels.api as mysm
from sklearn.linear_model import LogisticRegression
import numpy as mynp
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn import svm

In [2]:
#Import the dataset
mydata = mypd.read_excel('/Users/sudiptarakshit/Desktop/ISI MS QMS/Sem2/MDA/Assignment/Assignment8/Heart_Disease_Data.xlsx')

In [3]:
#Explore the dataset
mydata.head()

Age  Sex  CP  RestBP  Cholesteral  FBP  RestECG  Max_HR  ExAngina  Oldpeak  \
0   63    1   3     145          233    1        0     150         0      2.3   
1   37    1   2     130          250    0        1     187         0      3.5   
2   41    0   1     130          204    0        0     172         0      1.4   
3   56    1   1     120          236    0        1     178         0      0.8   
4   57    0   0     120          354    0        1     163         1      0.6   

   Slope  CA  Thal  Result  
0      0   0     1       1  
1      0   0     2       1  
2      2   0     2       1  
3      2   0     2       1  
4      2   0     2       1

In [4]:
#Check the dataset
mydata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          303 non-null    int64  
 1   Sex          303 non-null    int64  
 2   CP           303 non-null    int64  
 3   RestBP       303 non-null    int64  
 4   Cholesteral  303 non-null    int64  
 5   FBP          303 non-null    int64  
 6   RestECG      303 non-null    int64  
 7   Max_HR       303 non-null    int64  
 8   ExAngina     303 non-null    int64  
 9   Oldpeak      303 non-null    float64
 10  Slope        303 non-null    int64  
 11  CA           303 non-null    int64  
 12  Thal         303 non-null    int64  
 13  Result       303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


We find that there are total 303 entries in the dataset and there is no missing value present.

There are 13 expalnatory features in the data. 

All the features are numeric type. Some of the features are ordinal type discrete numeric data apparently.

In [5]:
#Separate x's and y
x = mydata.iloc[:,0:13]
y = mydata.iloc[:,13]

## Split the data randomly into training (80%) and test (20%)

We initialize the random number generator.

In [6]:
#Set seed
random.seed(100)

In [7]:
#Split the data into training (80%) and testing (20%)
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=100)

In [8]:
#Check the test/train split
x_train.count(), x_test.count()

(Age            242
 Sex            242
 CP             242
 RestBP         242
 Cholesteral    242
 FBP            242
 RestECG        242
 Max_HR         242
 ExAngina       242
 Oldpeak        242
 Slope          242
 CA             242
 Thal           242
 dtype: int64,
 Age            61
 Sex            61
 CP             61
 RestBP         61
 Cholesteral    61
 FBP            61
 RestECG        61
 Max_HR         61
 ExAngina       61
 Oldpeak        61
 Slope          61
 CA             61
 Thal           61
 dtype: int64)

train data has 242 observations & test data has 61 observations.

| | **Operation on Train data** |  |
| :----: | :----: | :----: |
| | Build a General Model |  |
| |⇩| |
| |R<sup>2</sup> Accuracy Check | |
| |⇩| |
| |Cross Validation R<sup>2</sup> Check for generalizability | |
| |⇩ | |
| |Hyper Parameter Tuning| |
| |⇩ | |
| |Build model using the best parameters| |
| |⇩ | |
| |R<sup>2</sup> Accuracy Check | |
| |⇩ | |
| |Cross Validation R<sup>2</sup> Check for generalizability | |
| |⇩ | |
||Get predicted value on train data||
| |⇩ | |
|| Confusion Matrix||
| |⇩ | |
||Accuracy on Training Data||
| |⇩ | |
||Missclassification on training data||

| | **Operation on Test data** |  |
| :----: | :----: | :----: |
||Get predicted value on Test data||
| |⇩| |
|| Confusion Matrix||
| |⇩ | |
||Accuracy on Test Data||
| |⇩ | |
||Missclassification on Test data||

**Note:**
During hyper parameter tuning we do cross validation. So, after hyper parameter tuning we get the model with best generalizability.


## Naive Bayes

In [9]:
#Specify the model
mynaivemodel = GaussianNB()

In [10]:
#Fit the model to the data
mynaivemodel = mynaivemodel.fit(X=x_train,y=y_train)

In [11]:
#Accuracy of the model
accuracy = mynaivemodel.score(X=x_train,y=y_train)
accuracy*100

83.88429752066115

In [12]:
#Perform a 4 fold Cross Validation
mynaivescore = cross_val_score(estimator=mynaivemodel,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [13]:
#Cross Validation Accuracy
mynaivescore.mean()

0.7971994535519127

We see a some drop in model accuracy after cross validation. We have to do hyper-parameter tuning for a more accurate and more generalized model.

**Hyper parameter optimization**

In [14]:
#Hyper parameter tuning
mynaiveparamgrid = {'var_smoothing':mynp.logspace(start=0,stop=-9,num=10)}

In [15]:
# Set up search
mynaivesearch = GridSearchCV(estimator=mynaivemodel,param_grid=mynaiveparamgrid,scoring='accuracy',cv=5,verbose=True)

In [16]:
#Fit the search on training data
mynaivesearch = mynaivesearch.fit(X=x_train,y=y_train)

Fitting 5 folds for each of 10 candidates, totalling 50 fits


In [17]:
#Optimum parameter values
mynaivesearch.best_params_

{'var_smoothing': 0.0001}

In [18]:
#Specify optimum model
mynaivemodel = GaussianNB(var_smoothing=0.0001)

In [19]:
# Fit the model to training data
mynaivemodel = mynaivemodel.fit(X=x_train,y=y_train)

In [20]:
# Check the model accuracy
naiveaccuracy = mynaivemodel.score(x_train,y_train)
naiveaccuracy

0.8347107438016529

There is no improvement in model accuracy after hyper parameter optimization, But there should be increase in model generalizability.

In [21]:
#Cross validation accuracy
mynaivescore = cross_val_score(estimator=mynaivemodel,X=x_train,y=y_train,scoring='accuracy',cv=5)
mynaivescore.mean()

0.8305272108843538

We see that cross validation accuracy has increased. And hence model generalizability has improved.

In [22]:
# Predicted Values on train data
naivepredy = mynaivemodel.predict(X=x_train)

In [23]:
#Actual vs Predicted table on train data
mynaivetable = mypd.crosstab(y_train,naivepredy)
#Confusion matrix
mynaivetable

col_0    0    1
Result         
0       75   30
1       10  127

**Model accuracy from confusion matrix**

In [24]:
#True Positive
TP = 127
#False Positive
FP = 30
#True Negative
TN = 75
#False Negative
FN = 10

**Model accuracy**


In [25]:
#Accuracy
TP/(TP+FP)

0.8089171974522293

In [26]:
#Misclassification rate
1 - TP/(TP+FP)

0.19108280254777066

### Validate model on test data

In [27]:
#Get the predicted value on test data
naivepredtest = mynaivemodel.predict(X=x_test)

In [28]:
#Actual vs Predicted table on test data
mynaivetable = mypd.crosstab(y_test,naivepredtest)
#Confusion matrix
mynaivetable

col_0    0   1
Result        
0       27   6
1        2  26

**Model accuracy from confusion matrix of test data**

In [29]:
#True Positive
TP = 26
#False Positive
FP = 6
#True Negative
TN = 27
#False Negative
FN = 2

**Model accuracy**


In [30]:
#Accuracy
TP/(TP+FP)

0.8125

In [31]:
#Misclassification rate
1 - TP/(TP+FP)

0.1875

Model accuracy on test data = 80.89% which is similar to accuracy of 81.25% on train data.

| |Naive Bayes Classifier Model | |
| ----------- | ----------- |----------- |
| | Train Data |Test Data |
|Accuracy |80.89 |81.25|
|Misclassification rate |19.10|18.75|


## KNN Classifier

In [32]:
#Specify the model
myknnmodel = KNeighborsClassifier(n_neighbors=5)

In [33]:
#Fit the model to the data
myknnmodel = myknnmodel.fit(X=x_train,y=y_train)

In [34]:
#Model accuarcy
knnaccuracy = myknnmodel.score(x_train,y_train)
knnaccuracy

0.7520661157024794

In [35]:
#Perform a 4 fold Cross Validation
myknnscore = cross_val_score(estimator=myknnmodel,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [36]:
#Cross Validation Accuracy
myknnscore.mean()

0.6075136612021859

We see that model accuracy as well as model generalizability not good. We can do hyper parameter tuning.

**Hyper parameter tuning**

In [43]:
#Hyper parameter tuning
myknnparamgrid = {'n_neighbors':[3,4,5,6,7,8,9,10,11,12,13,14,15]}

In [44]:
# Set up search
myknnsearch = GridSearchCV(estimator=myknnmodel,param_grid=myknnparamgrid,scoring='accuracy',cv=5,verbose=True)

In [45]:
#Fit the search on training data
myknnsearch = myknnsearch.fit(X=x_train,y=y_train)

Fitting 5 folds for each of 13 candidates, totalling 65 fits


In [46]:
#Optimum parameter values
myknnsearch.best_params_

{'n_neighbors': 15}

In [47]:
#Specfy the model
myknnmodel = KNeighborsClassifier(n_neighbors=15)

In [48]:
#Fit the model
myknnmodel = myknnmodel.fit(X=x_train,y=y_train)

In [49]:
#Model accuarcy
knnaccuracy = myknnmodel.score(x_train,y_train)
knnaccuracy

0.6611570247933884

In [50]:
#Perform a 4 fold Cross Validation
myknnscore = cross_val_score(estimator=myknnmodel,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [51]:
#Cross Validation Accuracy
myknnscore.mean()

0.6487021857923497

In [52]:
# Predicted Values on train data
knnpredy = myknnmodel.predict(X=x_train)

In [53]:
#Actual vs Predicted table on train data
myknntable = mypd.crosstab(y_train,knnpredy)
#Confusion matrix
myknntable

col_0    0    1
Result         
0       50   55
1       27  110

**Model accuracy from confusion matrix of train data**

In [54]:
#True Positive
TP = 110
#False Positive
FP = 55
#True Negative
TN = 50
#False Negative
FN = 27

**Model accuracy**


In [55]:
#Accuracy
TP/(TP+FP)

0.6666666666666666

In [56]:
#Misclassification rate
1 - TP/(TP+FP)

0.33333333333333337

### Validate model on test data

In [57]:
#Get the predicted value on test data
knnpredtest = myknnmodel.predict(X=x_test)

In [58]:
#Actual vs Predicted table on test data
myknntable = mypd.crosstab(y_test,knnpredtest)
#Confusion matrix
myknntable

col_0    0   1
Result        
0       17  16
1        4  24

**Model accuracy from confusion matrix of test data**

In [59]:
#True Positive
TP = 24
#False Positive
FP =16
#True Negative
TN = 27
#False Negative
FN = 4

**Model accuracy**


In [60]:
#Accuracy
TP/(TP+FP)

0.6

In [61]:
#Misclassification rate
1 - TP/(TP+FP)

0.4

Model accuracy on test data = 79.41% which is sufficient drop from 93.70% on train data.

| |KNN Classifier Model | |
| ----------- | ----------- |----------- |
| | Train Data |Test Data |
|Accuracy |66.66 |60.00|
|Misclassification rate |33.33|40.00|


## SVM Classifier Method

In [62]:
#Specify the model
mysvmmodel = svm.SVC()

In [63]:
#Fit the model to training data
mysvmmodel = mysvmmodel.fit(x_train,y_train)

In [64]:
#Model accuarcy
svmaccuracy = mysvmmodel.score(x_train,y_train)
round(svmaccuracy,3)

0.649

Model accuracy is not good. We check model generalizability using cross validation.

In [65]:
#Perform a 4 fold Cross Validation
mysvmscore = cross_val_score(estimator=mysvmmodel,X=x_train,y=y_train,cv=4,scoring='accuracy')

In [66]:
#Cross Validation Accuracy
mysvmscore.mean()

0.6323770491803279

We see that model generalizability is good with train data.

**Hyper parameter optimization**

In [67]:
#Specify the parameter grid
mysvmparamgrid = {'kernel':['linear', 'poly', 'rbf'],'degree':[1,2,3]}

In [68]:
# Set up search
mysvmsearch = GridSearchCV(estimator=mysvmmodel,param_grid=mysvmparamgrid,scoring='accuracy',cv=5,verbose=True)

In [69]:
#Fit the search on training data
mysvmsearch = mysvmsearch.fit(X=x_train,y=y_train)

Fitting 5 folds for each of 9 candidates, totalling 45 fits


In [70]:
#Optimum parameter values
mysvmsearch.best_params_

{'degree': 1, 'kernel': 'linear'}

In [71]:
#Specify optimum model
mysvmmodel = svm.SVC(kernel='linear',degree=1)

In [72]:
# Fit the model to training data
mysvmmodel = mysvmmodel.fit(X=x_train,y=y_train)

In [73]:
# Check the model accuracy:
svmaccuracy = mysvmmodel.score(x_train,y_train)
svmaccuracy

0.8636363636363636

In [74]:
#Cross validation accuracy
mysvmscore = cross_val_score(estimator=mysvmmodel,X=x_train,y=y_train,scoring='accuracy',cv=5)
mysvmscore.mean()

0.8306122448979592

In [75]:
# Predicted Values on train data
svmpredy = mysvmmodel.predict(X=x_train)

In [76]:
#Actual vs Predicted table on train data
mysvmtable = mypd.crosstab(y_train,svmpredy)
#Confusion matrix
mysvmtable

col_0    0    1
Result         
0       83   22
1       11  126

**Model accuracy from confusion matrix**

In [77]:
#True Positive
TP = 126
#False Positive
FP = 22
#True Negative
TN = 83
#False Negative
FN = 11

**Model accuracy**


In [78]:
#Accuracy
TP/(TP+FP)

0.8513513513513513

In [79]:
#Misclassification rate
1 - TP/(TP+FP)

0.14864864864864868

### Validate model on test data

In [80]:
#Get the predicted value on test data
svmpredtest = mysvmmodel.predict(X=x_test)

In [81]:
#Actual vs Predicted table on test data
mysvmtable = mypd.crosstab(y_test,svmpredtest)
#Confusion matrix
mysvmtable

col_0    0   1
Result        
0       24   9
1        4  24

**Model accuracy from confusion matrix of test data**

In [82]:
#True Positive
TP = 24
#False Positive
FP = 9
#True Negative
TN = 24
#False Negative
FN = 4

**Model accuracy**


In [83]:
#Accuracy
TP/(TP+FP)

0.7272727272727273

In [84]:
#Misclassification rate
1 - TP/(TP+FP)

0.2727272727272727

Model accuracy on test data = 72.72% which is sufficient drop from 85.13% on train data.

| |SVM Classifer | |
| ----------- | ----------- |----------- |
| | Train Data |Test Data |
|Accuracy |85.13 |72.72|
|Misclassification rate |14.86|27.27|


**We comapare all three models**

| |Naive Bayes |  |KNN Classifier| |SVM Classifier| |
| :-----------:| :-----------: |:-----------: | :-----------: |:-----------:|:-----------:|-----------:|
| | Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |
| Accuracy |80.89|81.25|66.66|60.00|85.13|72.72|
|Misclassification rate |19.10 |18.75|33.33 |40.00|14.86|27.27|



After comparing KNN classifier, SVM classifier and Naive Bayes we see that KNN Classifier has the lowest accuracy, whereas SVM Classifier has highest accuracy.

## Compare all the models

| |Bagging Model |  |Random Forest| |CART| |Logistic Regression| |Naive Bayes| |KNN| |SVM| |
| :-----------: | :-----------: |:-----------: | :-----------: |:-----------:|:-----------: |:-----------: |:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|:-----------:|
| | Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |Train Data |Test Data |
| Accuracy |93.70|79.41|97.82|78.12|91.15|71.42|84.3|83.6|80.89|81.25|66.66|60.00|85.13|72.27|
|Misclassification Rate |6.29 |20.59|2.17 |21.87|8.84|28.57|15.7|16.4|19.10|18.75|33.33|40.00|14.86|27.27|



Comapring all the models we see that Tree based models are giving higher accuarcy where as KNN classifier is giving the least accuracy.